### PR TITLE
False positive ``unsubscriptable-object``

### LINKED ISSUE #2305: Regression in Astroid version 2.15.7 in handling subscriptable type parameters
Astroid version 2.15.7 fails to correctly handle a subscriptable type parameter  most likely due to the change in this [PR](https://github.com/pylint-dev/astroid/pull/2239). 

### Steps to reproduce

```python
from collections.abc import Mapping
from typing import Generic, TypeVar, TypedDict
from dataclasses import dataclass

class Identity(TypedDict):
    """It's the identity."""

    name: str

T = TypeVar("T", bound=Mapping)

@dataclass
class Animal(Generic[T]):
    """It's an animal."""

    identity: T

class Dog(Animal[Identity]):
    """It's a Dog."""

dog = Dog(identity=Identity(name="Dog"))
print(dog.identity["name"])
```

### Current behavior
Pylint (running Astroid 2.15.7) gives the following error for the example above:
```
E1136: Value 'dog.identity' is unsubscriptable (unsubscriptable-object)
```
### Expected behavior
Astroid should correctly handle a subscriptable type parameter.

### `python -c "from astroid import __pkginfo__; print(__pkginfo__.version)"` output
2.15.7

### PR DESCRIPTION
**Pull request by mbyrnepr2**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
|    | :sparkles: New feature |
|  ✓  | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description
- Fix a false positive for `unsubscriptable-object` (regression in 2.15.7).
- Refactor to raise an `InferenceError` if the `TypeVar` naming is invalid.

Closes #2305

### CODE REVIEW

**Comment by mbyrnepr2** [comment]
I'm thinking of adding the test examples from this issue and https://github.com/pylint-dev/pylint/issues/9069 to the Pylint functional tests.

File: astroid/brain/brain_typing.py [hunk 1]
```diff
@@ -118,9 +119,7 @@ def looks_like_typing_typevar_or_newtype(node) -> bool:
     return False
 
 
-def infer_typing_typevar_or_newtype(
-    node: Call, context_itton: context.InferenceContext | None = None
-) -> Iterator[ClassDef]:
+def infer_typing_typevar_or_newtype(node, context_itton=None):
```
Comment (Pierre-Sassoulas) [inline astroid/brain/brain_typing.py:122]:
```suggestion
def infer_typing_typevar_or_newtype(
    node: Call, context_itton: context.InferenceContext | None = None
) -> Iterator[ClassDef]:
```
We could keep the typing, right ?

Comment (mbyrnepr2) [reply astroid/brain/brain_typing.py]: Well, yes :)

File: tests/brain/test_typing.py [hunk 2]
```diff
@@ -15,10 +18,8 @@ def test_infer_typevar() -> None:
...
-    MyType = TypeVar('My.Type')
+    TypeVar('My.Type')
     """
     )
-    call = assign_node.value
-    inferred = next(call.infer())
-    assert isinstance(inferred, nodes.ClassDef)
-    assert inferred.name == "My.Type"
+    with pytest.raises(InferenceError):
+        assign_node.inferred()
```
Comment (jacobtylerwalls) [inline tests/brain/test_typing.py:25]: Small nit, but it's no longer an "assign node" given the change to the parsed code.

Comment (mbyrnepr2) [reply tests/brain/test_typing.py]: Cheers!

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the quick work!

**Pull request merged**